server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java

```diff
@@ -136,52 +129,97 @@ protected void masterOperation(
-        // construction upstream when instantiating these
-        private final class CreateIndexTask extends AckedClusterStateUpdateTask {
-
-            final CreateIndexRequest request;
-            final AtomicReference<String> indexNameRef;
+        private final class CreateIndexTask implements ClusterStateTaskListener {
+            private final CreateIndexRequest request;
+            private final ActionListener<CreateIndexResponse> listener;
 
-            CreateIndexTask(
-                CreateIndexRequest request,
-                ActionListener<AcknowledgedResponse> listener,
-                AtomicReference<String> indexNameRef
-            ) {
-                super(Priority.URGENT, request, listener);
+            private CreateIndexTask(CreateIndexRequest request, ActionListener<CreateIndexResponse> listener) {
                 this.request = request;
-                this.indexNameRef = indexNameRef;
+                this.listener = listener;
             }
 
             @Override
-            public ClusterState execute(ClusterState currentState) throws Exception {
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                assert false : "should not be called";
+            }
+
+            private ClusterStateAckListener getAckListener(String indexName) {
+                return new ClusterStateAckListener() {
+                    @Override
+                    public boolean mustAck(DiscoveryNode discoveryNode) {
+                        return true;
+                    }
+
+                    @Override
+                    public void onAllNodesAcked(Exception e) {
+                        if (e == null) {
+                            activeShardsObserver.waitForActiveShards(
+                                new String[] { indexName },
+                                ActiveShardCount.DEFAULT,
+                                request.timeout(),
+                                shardsAcked -> listener.onResponse(new CreateIndexResponse(true, shardsAcked, indexName)),
+                                listener::onFailure
+                            );
+                        } else {
+                            listener.onResponse(new CreateIndexResponse(false, false, indexName));
+                        }
+                    }
+
+                    @Override
+                    public void onAckTimeout() {
+                        listener.onResponse(new CreateIndexResponse(false, false, indexName));
+                    }
+
+                    @Override
+                    public TimeValue ackTimeout() {
+                        return request.ackTimeout();
+                    }
+                };
+            }
+
+            ClusterState execute(
```

Comment:
Is it worth a comment on the implied contract w.r.t. returning a clusterState, calling success/failure on the taskContext, and updating successfulRequests? It looks to me like you're doing all the juggling just fine, but some future developer might be happy to add a new return case and think that just satisfying the compiler here is sufficient.

Reply:
Yes, good point. I added a comment in 6be0514 and also an assertion to enforce it.